### PR TITLE
Expose enabled and active states on interaction handlers

### LINKED ISSUE #2173: Expose interaction handler enabled/disabled state
There's an issue on mapbox-gl-js-draw asking for disabling boxZoom temporarily when selecting features (https://github.com/mapbox/mapbox-gl-draw/issues/213), however, as far as I can tell there is no supported way to determine whether a certain interaction handler is enabled or not.

Even if it would be possible to always disable boxZoom temporarily when selecting features, to determine whether to enable it again requires knowledge of whether boxZoom was intentionally disabled overall.

Would it make sense to expand the API to expose the state of interaction handlers?

### PR DESCRIPTION
**Pull request by averas**
Fixes #2173.

While working on this I found this anomaly: #2237, please let me know if you think that one should be tracked down and addressed as part of this PR.

### CODE REVIEW

File: js/ui/handler/box_zoom.js [hunk 1]
```diff
@@ -21,6 +23,9 @@ function BoxZoomHandler(map) {
 
 BoxZoomHandler.prototype = {
 
+    enabled: false,
+    active: false,
```
Comment (lucaswoj) [inline js/ui/handler/box_zoom.js]: Are you using `active` anywhere?

Comment (averas) [inline js/ui/handler/box_zoom.js]: Some of the handlers, such as  <code>BoxZoomHandler</code>, already had a public (no underscore), albeit non-documented <code>active</code> property on them, I just took the opportunity to make code and behaviour consistent with the new enabled property.

Comment (lucaswoj) [inline js/ui/handler/box_zoom.js]: Got it! Thanks :smile:

File: js/ui/handler/keyboard.js [hunk 2]
```diff
@@ -20,6 +20,7 @@ var panDelta = 80,
  *  * `Shift+⇡`: increase pitch by 5 degrees
  *  * `Shift+⇣`: decrease pitch by 5 degrees
  * @class KeyboardHandler
+ * @property {boolean} enabled Whether the keyboard interaction is currently enabled
```
Comment (lucaswoj) [inline js/ui/handler/keyboard.js]: > Whether ~~the~~ keyboard interaction is currently enabled

**Comment by lucaswoj** [comment]
Looks great! Let's give @bhousel a chance to look this over and then we can merge it.

**Comment by averas** [comment]
Btw. @lucaswoj. There was a discussion on #2085 related to this. Obviously now that the handler has knowledge of whether it in fact is enabled or not the fallback of always disabling a handler before enabling it could be removed by just making <code>enable()</code> and <code>disable()</code> noops if the desired state is already present.

Let me know if you prefer this refactoring in contrast to the current fallback behaviour.

**Comment by bhousel** [comment]
A problem with the previous code, but `active` should probably be an accessor not a public property.  For example in the dragpan handler, somebody could start panning, and if code elsewhere sets that public `active=false`, the handler will never run the things in `_onUp()`.

It's necessary for some handlers to know which other handlers are active, but no code outside the handler should be allowed to modify that state.

**Comment by averas** [comment]
@bhousel That's true, and obviously an inherent problem with mutable properties. 

With the same reasoning I take it you are for keeping the <code>disable()</code> fallback call in the <code>enable()</code> functions, rather than making them noops based on current state (as outlined in my previous comment). Otherwise you may run into the same problem with the enable property, i.e. the property being mutated directly bringing the handler into a degenerative state.

**Comment by lucaswoj** [comment]
# Next Steps

 - [ ] prefix `_active` and `_enabled` with underscores to mark them as private properties
 - [ ] add and document public `isActive` and `isEnabled` accessor methods
 - [ ] add a brief documentation note about the difference between "active" and "enabled"

Anything else we should add to this list @bhousel?

**Comment by averas** [comment]
@lucaswoj The reason I went down the path of public properties in favour of accessors was @mourner's decision in #2173. Of course we can change that in accordance with your outline. Your suggested changes will be breaking, but I highly doubt anybody is currently using the non-documented <code>active</code> property.

**Comment by averas** [comment]
@lucaswoj np. I'll provide a new set of commits...

**Comment by averas** [comment]
Changes made according to @lucaswoj outline.

:eyes: @lucaswoj @bhousel

**Comment by lucaswoj** [comment]
Thank you @averas! I'm going to test all the interaction handlers by hand and then I feel good about :ship: this!

 - [ ] KeyboardHandler
 - [ ] DragPanHandler
 - [ ] TouchZoomRotateHandler
 - [ ] ScrollZoomHandler
 - [ ] DoubleClickZoomHandler
 - [ ] DragRotateHandler

**Comment by averas** [comment]
@lucaswoj Great! I made some additional comments on #2237 on that issue.

**Comment by lucaswoj** [comment]
Merged in 494e2bb52be71d69d3a8f12904d3186824fb99a3! Thanks again @averas! :ship: :shipit: :rocket:

**Pull request closed without merging**